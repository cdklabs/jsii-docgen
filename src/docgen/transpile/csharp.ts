import * as Case from 'case';
import * as reflect from 'jsii-reflect';
import * as transpile from './transpile';

export class CSharpTranspile extends transpile.TranspileBase {
  constructor() {
    super(transpile.Language.CSHARP);
  }

  public moduleLike(
    moduleLike: reflect.ModuleLike,
  ): transpile.TranspiledModuleLike {
    const csharpPackage: string = moduleLike.targets?.dotnet?.namespace;

    // if this is a submodule, we need to break the package name down into the
    // parent name and the submodule. we also allow submodules not to have
    // explicit target names, in which case we need to append the pascal-cased
    // submodule name to the parent package name.
    if (moduleLike instanceof reflect.Submodule) {
      const parent = this.getParentModule(moduleLike);
      const parentFqn = parent.targets?.dotnet?.namespace;

      // if the submodule does not explicitly define a dotnet package name, we need to deduce it from the parent
      const submoduleCSharpPackage = csharpPackage ?? `${parentFqn}.${Case.pascal(moduleLike.name)}`;
      return { name: parentFqn, submodule: submoduleCSharpPackage };
    }

    return { name: csharpPackage };
  }

  public type(type: reflect.Type): transpile.TranspiledType {
    const submodule = this.findSubmodule(type);
    const moduleLike = this.moduleLike(submodule ? submodule : type.assembly);

    const fqn = [];

    let namespace = type.namespace;
    if (namespace) {
      fqn.push(moduleLike.name);
      fqn.push(namespace);
    } else {
      fqn.push(moduleLike.name);
    }
    fqn.push(type.name);

    return {
      fqn: fqn.join('.'),
      name: type.name,
      namespace: namespace,
      module: moduleLike.name,
      submodule: moduleLike.submodule,
      source: type,
      language: this.language,
    };
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);
    const isInitializer = callable.kind === reflect.MemberKind.Initializer;
    const name = isInitializer
      ? type.name
      : Case.pascal(callable.name);

    const params = callable.parameters.map(p => this.formatParameter(this.parameter(p))).join(', ');
    const signatures = [`public ${name}(${params})`];
    const invocations = [isInitializer
      ? `new ${name}(${params});`
      : `${type.name}.${name}(${params});`];

    return {
      name,
      parentType: type,
      import: this.formatImport(type),
      parameters: callable.parameters,
      signatures,
      invocations,
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    const indent = ' '.repeat(4);
    const inputs = struct.allProperties.map((p) =>
      this.formatBuilderMethod(this.property(p), indent),
    ).flat();
    return {
      type: type,
      name: struct.name,
      import: this.formatImport(type),
      initialization: this.formatStructBuilder(type, inputs),
    };
  }

  public interface(
    iface: reflect.InterfaceType,
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  public parameter(
    parameter: reflect.Parameter,
  ): transpile.TranspiledParameter {
    const typeRef = this.typeReference(parameter.type);
    return {
      name: parameter.name,
      parentType: this.type(parameter.parentType),
      typeReference: typeRef,
      optional: parameter.optional,
      declaration: this.formatProperty(parameter.name, typeRef),
    };
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    const typeRef = this.typeReference(property.type);
    return {
      name: property.name,
      parentType: this.type(property.parentType),
      typeReference: typeRef,
      optional: property.optional,
      declaration: this.formatProperty(property.name, typeRef),
    };
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }

  public unionOf(): string {
    return 'object';
  }

  public listOf(type: string): string {
    return `${type}[]`;
  }

  public mapOf(type: string): string {
    return `System.Collections.Generic.IDictionary<string, ${type}>`;
  }

  public any(): string {
    return 'object';
  }

  public str(): string {
    return 'string';
  }

  public number(): string {
    return 'double';
  }

  public boolean(): string {
    return 'bool';
  }

  public json(): string {
    return 'Newtonsoft.Json.Linq.JObject';
  }

  public date(): string {
    return 'System.DateTime';
  }

  public readme(readme: string): string {
    return readme;
  }

  private formatImport(type: transpile.TranspiledType): string {
    return `using ${type.module};`;
  };

  private formatParameter(
    transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty,
  ): string {
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${tf} ${transpiled.name}`;
  }

  private formatStructBuilder(type: transpile.TranspiledType, properties: string[]): string {
    const builder = `new ${type.name} {`;
    return [
      builder,
      ...properties,
      '};',
    ].join('\n');
  };

  private formatBuilderMethod(
    transpiled: transpile.TranspiledParameter,
    indent: string,
  ): string[] {
    return [`${indent}${transpiled.name}`];
  }

  private formatProperty(
    name: string,
    typeReference: transpile.TranspiledTypeReference,
  ): string {
    const tf = typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${tf} ${name}`;
  }
}
