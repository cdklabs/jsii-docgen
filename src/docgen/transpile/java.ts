import * as reflect from 'jsii-reflect';
import * as transpile from './transpile';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const Case = require('case');

// Helper methods
const toCamelCase = (text?: string) => {
  return Case.camel(text ?? '');
};

const toUpperCamelCase = (test?: string) => {
  const camelCase = toCamelCase(test);
  return camelCase.charAt(0).toUpperCase() + camelCase.slice(1);
};

// [1, 2, 3] -> [[], [1], [1, 2], [1, 2, 3]]
const prefixArrays = <T>(arr: T[]): T[][] => {
  const out: T[][] = [[]];
  const prefix: T[] = [];
  for (const elem of arr) {
    prefix.push(elem);
    out.push([...prefix]);
  }
  return out;
};

/**
 * Hack to convert a jsii property to a parameter for
 * parameter expansion.
 */
const propertyToParameter = (
  callable: reflect.Callable,
  property: reflect.Property,
): reflect.Parameter => {
  return {
    docs: property.docs,
    method: callable,
    name: property.name,
    optional: property.optional,
    parentType: property.parentType,
    spec: property.spec,
    system: property.system,
    type: property.type,
    variadic: false,
  };
};

export class JavaTranspile extends transpile.TranspileBase {
  constructor() {
    super(transpile.Language.JAVA);
  }

  public moduleLike(
    moduleLike: reflect.ModuleLike,
  ): transpile.TranspiledModuleLike {
    const fqn: string = moduleLike.targets?.java?.package;
    if (!fqn) {
      throw new Error(
        `Java is not a supported target for module: ${moduleLike.fqn}`,
      );
    }

    if (moduleLike instanceof reflect.Submodule) {
      const parent = this.getParentModule(moduleLike);
      if (!parent) {
        throw new Error(`Could not find parent module of ${moduleLike.fqn}`);
      }
      const parentFqn = parent.targets?.java?.package;
      if (!fqn.startsWith(parentFqn)) {
        throw new Error(`Expected submodule ${fqn} to start with ${parentFqn} since it is its parent module.`);
      }
      // { name: "software.amazon.awscdk", submodule: "services.ecr" }
      return { name: parentFqn, submodule: fqn.substring(parentFqn.length + 1) };
    }

    return { name: fqn };
  }

  private getParentModule(moduleLike: reflect.ModuleLike): reflect.Assembly | undefined {
    const ts = moduleLike.system;
    for (const assembly of ts.assemblies) {
      const child = assembly.submodules.find((mod) => mod.fqn === moduleLike.fqn);
      if (child) {
        return assembly;
      }
    }
    return undefined;
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);

    const parameters = callable.parameters.sort(this.optionalityCompare);

    const requiredParams = parameters.filter((p) => !p.optional);
    const optionalParams = parameters.filter((p) => p.optional);

    const name = callable.name;

    // simulate Java method overloading
    const inputLists = prefixArrays(optionalParams).map((optionals) => {
      return [...requiredParams, ...optionals].map((p) => this.formatParameter(this.parameter(p)));
    });

    const signatures = inputLists.map((inputs) => {
      return this.formatSignature(name, inputs);
    });

    let invocations;

    if (this.isClassBuilderGenerated(callable)) {
      const struct = this.extractFirstStruct(parameters);

      // render using Java builder syntax (show no overloads)
      invocations = [this.formatClassBuilder(type, parameters, struct)];

      // flatten out the parameters so the user doesn't have to jump between
      // docs of Foo and FooProps
      for (const property of struct.allProperties) {
        const parameter = propertyToParameter(callable, property);
        parameters.push(parameter);
      }
    } else {
      invocations = callable.kind === reflect.MemberKind.Initializer
        // render with `new Class` syntax (showing all constructor overloads)
        ? inputLists.map((inputs) => this.formatClassInitialization(type, inputs))
        // render invocation as method calls (showing all method overloads)
        : inputLists.map((inputs) => this.formatInvocation(type, inputs, name));
    }

    return {
      name,
      parentType: type,
      import: this.formatImport(type),
      parameters,
      signatures,
      invocations,
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    const indent = ' '.repeat(4);
    const inputs = struct.allProperties.map((p) =>
      this.formatBuilderMethod(this.property(p), indent),
    ).flat();
    return {
      type: type,
      name: struct.name,
      import: this.formatImport(type),
      initialization: this.formatStructBuilder(type, inputs),
    };
  }

  public interface(
    iface: reflect.InterfaceType,
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  public parameter(
    parameter: reflect.Parameter,
  ): transpile.TranspiledParameter {
    const typeRef = this.typeReference(parameter.type);
    return {
      name: parameter.name,
      parentType: this.type(parameter.parentType),
      typeReference: typeRef,
      optional: parameter.optional,
      declaration: this.formatProperty(parameter.name, typeRef),
    };
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    const typeRef = this.typeReference(property.type);
    return {
      name: property.name,
      parentType: this.type(property.parentType),
      typeReference: typeRef,
      optional: property.optional,
      declaration: this.formatProperty(property.name, typeRef),
    };
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }

  public unionOf(types: string[]): string {
    return types.join(' OR ');
  }

  public listOf(type: string): string {
    return `java.util.List<${type}>`;
  }

  public mapOf(type: string): string {
    return `java.util.Map<java.lang.String, ${type}>`;
  }

  public any(): string {
    return 'java.lang.Object';
  }

  public str(): string {
    return 'java.lang.String';
  }

  public number(): string {
    return 'java.lang.Number';
  }

  public boolean(): string {
    return 'java.lang.Boolean';
  }

  public json(): string {
    return 'com.fasterxml.jackson.databind.node.ObjectNode';
  }

  public date(): string {
    return 'java.time.Instant';
  }

  public readme(readme: string): string {
    return readme;
  }

  private formatImport(type: transpile.TranspiledType): string {
    return `import ${type.fqn};`;
  };

  private formatParameter(
    transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty,
  ): string {
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${tf} ${transpiled.name}`;
  }

  private formatInputs(inputs: string[]): string {
    return inputs.join(', ');
  };

  private formatStructBuilder(type: transpile.TranspiledType, methods: string[]): string {
    const builder = `${type.name}.builder()`;
    return [
      builder,
      ...methods,
      '    .build();',
    ].join('\n');
  };

  private formatClassInitialization(
    type: transpile.TranspiledType,
    inputs: string[],
  ): string {
    return `new ${type.name}(${this.formatInputs(inputs)});`;
  };

  private formatClassBuilder(
    type: transpile.TranspiledType,
    parameters: reflect.Parameter[],
    struct: reflect.InterfaceType,
  ): string {
    const createArgs = this.formatInputs(parameters.map((p) => this.formatParameter(this.parameter(p))));
    const indent = ' '.repeat(4);
    const methods: string[] = struct.allProperties.map((p) =>
      this.formatBuilderMethod(this.property(p), indent),
    ).flat();
    return [
      `${type.name}.Builder.create(${createArgs})`,
      ...methods,
      `${indent}.build();`,
    ].join('\n');
  };


  private formatSignature(name: string, inputs: string[]) {
    return `public ${name}(${this.formatInputs(inputs)})`;
  };

  private formatBuilderMethod(
    transpiled: transpile.TranspiledParameter,
    indent: string,
  ): string[] {
    if (transpiled.optional) indent = '//' + indent.slice(2);
    const lowerCamel = toCamelCase(transpiled.name);
    const base = `${indent}.${lowerCamel}`;
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    // allow rendering union types as multiple overrided builder methods
    if (tf.includes(' OR ')) {
      const choices = tf.split(' OR ');
      return choices.map((typ) => `${base}(${typ})`);
    } else {
      return [`${base}(${tf})`];
    }
  }

  private formatInvocation(
    type: transpile.TranspiledType,
    inputs: string[],
    method: string,
  ): string {
    let target = type.submodule ? `${type.namespace}.${type.name}` : type.name;
    if (method) {
      target = `${target}.${method}`;
    }
    return `${target}(${this.formatInputs(inputs)})`;
  };

  private optionalityCompare(
    p1: reflect.Parameter,
    p2: reflect.Parameter,
  ): number {
    if (!p1.optional && p2.optional) {
      return -1;
    }
    if (!p2.optional && p1.optional) {
      return 1;
    }
    return 0;
  }

  private isStruct(p: reflect.Parameter): boolean {
    return p.type.fqn ? p.system.findFqn(p.type.fqn).isDataType() : false;
  }

  private isClassBuilderGenerated(
    callable: reflect.Callable,
  ): boolean {
    if (callable.kind !== reflect.MemberKind.Initializer) {
      return false;
    }

    const parameters: reflect.Parameter[] = callable.parameters.sort(this.optionalityCompare);
    const firstStruct = parameters.find((param) => this.isStruct(param));

    // no builder is generated if there is no struct parameter
    if (!firstStruct) {
      return false;
    }

    return true;
  };

  /**
   * Extracts the first struct out of a list of parameters (and throws
   * if there is none), removing it from the array.
   */
  private extractFirstStruct(
    parameters: reflect.Parameter[],
  ): reflect.InterfaceType {
    const firstStruct = parameters.find((param) => this.isStruct(param));
    if (!firstStruct) {
      throw new Error('No struct found in parameter list.');
    }
    const struct = firstStruct.parentType.system.findInterface(firstStruct.type.fqn!);
    parameters.splice(parameters.indexOf(firstStruct), 1);
    return struct;
  }

  private formatProperty(
    name: string,
    typeReference: transpile.TranspiledTypeReference,
  ): string {
    const tf = typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    if (tf.includes(' OR ')) {
      return `public java.lang.Object get${toUpperCamelCase(name)}();`;
    } else {
      return `public ${tf} get${toUpperCamelCase(name)}();`;
    }
  }
}
