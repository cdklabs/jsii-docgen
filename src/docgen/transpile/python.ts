import * as reflect from 'jsii-reflect';
import * as transpile from './transpile';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const Case = require('case');

// Helpers
const toSnakeCase = (text?: string) => {
  return Case.snake(text ?? '');
};

const formatImport = (type: transpile.TranspiledType) => {
  if (type.submodule) {
    return `from ${type.module} import ${type.submodule}`;
  }
  return `import ${type.module}`;
};

const formatArguments = (inputs: string[]) => {
  return inputs.length === 0 ? '()' : [
    '(',
    inputs.map(i => `  ${i}`).join(',\n'),
    ')',
  ].join('\n');
};

const formatInvocation = (
  type: transpile.TranspiledType,
  inputs: string[],
  method?: string,
) => {
  let target;
  if (type.submodule) {
    if (!type.namespace) {
      throw new Error(
        `Invalid type: ${type.fqn}: Types defined in a submodule (${type.submodule}) must have a namespace. `,
      );
    }
    // we don't include the submodule name here since it is
    // included in the namespace. this works because we import the submodule
    // in this case.
    // TODO - merge `formatInvocation` with `formatImport` since they are inherently coupled.
    target = `${type.namespace}.${type.name}`;
  } else {
    target = type.fqn;
  }

  if (method) {
    target = `${target}.${method}`;
  }
  return `${target}${formatArguments(inputs)}`;
};

const formatSignature = (name: string, inputs: string[]) => {
  const def = 'def ';
  return `${def}${name}${formatArguments(inputs)}`;
};

/**
 * Hack to convert a jsii property to a parameter for
 * python specific parameter expansion.
 */
const propertyToParameter = (
  callable: reflect.Callable,
  property: reflect.Property,
): reflect.Parameter => {
  return {
    docs: property.docs,
    method: callable,
    name: property.name,
    optional: property.optional,
    parentType: property.parentType,
    spec: property.spec,
    system: property.system,
    type: property.type,
    variadic: false,
  };
};

/**
 * A python transpiler.
 */
export class PythonTranspile extends transpile.TranspileBase {
  constructor() {
    super(transpile.Language.PYTHON);
  }

  public readme(readme: string): string {
    return readme;
  }

  public unionOf(types: string[]): string {
    return `${this.typing('Union')}[${types.join(', ')}]`;
  }

  public listOf(type: string): string {
    return `${this.typing('List')}[${type}]`;
  }

  public mapOf(type: string): string {
    return `${this.typing('Mapping')}[${type}]`;
  }

  public any(): string {
    return this.typing('Any');
  }

  public boolean(): string {
    return 'bool';
  }

  public str(): string {
    return 'str';
  }

  public number(): string {
    return `${this.typing('Union')}[int, float]`;
  }

  public date(): string {
    return 'datetime.datetime';
  }

  public enum(enu: reflect.EnumType): transpile.TranspiledEnum {
    return {
      fqn: this.type(enu).fqn,
      name: enu.name,
    };
  }

  public enumMember(em: reflect.EnumMember): transpile.TranspiledEnumMember {
    return {
      fqn: `${this.enum(em.enumType).fqn}.${em.name}`,
      name: em.name,
    };
  }
  public json(): string {
    return 'any';
  }

  public property(property: reflect.Property): transpile.TranspiledProperty {
    const name = property.const ? property.name : toSnakeCase(property.name);
    const typeRef = this.typeReference(property.type);
    return {
      name,
      parentType: this.type(property.parentType),
      typeReference: typeRef,
      optional: property.optional,
      declaration: this.formatProperty(name, typeRef),
    };
  }

  public class(klass: reflect.ClassType): transpile.TranspiledClass {
    return {
      name: klass.name,
      type: this.type(klass),
    };
  }

  public parameter(
    parameter: reflect.Parameter,
  ): transpile.TranspiledParameter {
    const name = toSnakeCase(parameter.name);
    const typeRef = this.typeReference(parameter.type);
    return {
      name,
      parentType: this.type(parameter.parentType),
      typeReference: typeRef,
      optional: parameter.optional,
      declaration: this.formatProperty(name, typeRef),
    };
  }

  public struct(struct: reflect.InterfaceType): transpile.TranspiledStruct {
    const type = this.type(struct);
    const inputs = struct.allProperties.map((p) =>
      this.formatParameters(this.property(p)),
    );
    return {
      type: type,
      name: struct.name,
      import: formatImport(type),
      initialization: formatInvocation(type, inputs),
    };
  }

  public callable(callable: reflect.Callable): transpile.TranspiledCallable {
    const type = this.type(callable.parentType);

    const parameters = new Array<reflect.Parameter>();

    for (const p of callable.parameters.sort(this.optionalityCompare)) {
      if (!this.isStruct(p)) {
        parameters.push(p);
      } else {
        // struct parameters are expanded to the individual struct properties
        const struct = p.parentType.system.findInterface(p.type.fqn!);
        for (const property of struct.allProperties) {
          const parameter = propertyToParameter(callable, property);
          parameters.push(parameter);
        }
      }
    }

    const name = toSnakeCase(callable.name);
    const inputs = parameters.map((p) => this.formatParameters(this.parameter(p)));

    return {
      name,
      parentType: type,
      import: formatImport(type),
      parameters,
      signatures: [formatSignature(name, inputs)],
      invocations: [formatInvocation(
        type,
        inputs,
        callable.kind === reflect.MemberKind.Initializer ? undefined : name,
      )],
    };
  }

  public moduleLike(
    moduleLike: reflect.ModuleLike,
  ): transpile.TranspiledModuleLike {
    const fqn = moduleLike.targets?.python?.module;
    if (!fqn) {
      throw new Error(
        `Python is not a supported target for module: ${moduleLike.fqn}`,
      );
    }

    if (moduleLike instanceof reflect.Submodule) {
      const fqnParts = fqn.split('.');
      return { name: fqnParts[0], submodule: fqnParts[1] };
    }
    return { name: fqn };
  }

  public interface(
    iface: reflect.InterfaceType,
  ): transpile.TranspiledInterface {
    return {
      name: iface.name,
      type: this.type(iface),
    };
  }

  private optionalityCompare(
    p1: reflect.Parameter,
    p2: reflect.Parameter,
  ): number {
    if (!p1.optional && p2.optional) {
      return -1;
    }
    if (!p2.optional && p1.optional) {
      return 1;
    }
    return 0;
  }

  private isStruct(p: reflect.Parameter): boolean {
    return p.type.fqn ? p.system.findFqn(p.type.fqn).isDataType() : false;
  }

  private typing(type: 'List' | 'Mapping' | 'Any' | 'Union'): string {
    return `typing.${type}`;
  }

  private formatParameters(
    transpiled: transpile.TranspiledParameter | transpile.TranspiledProperty,
  ): string {
    const tf = transpiled.typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${transpiled.name}: ${tf}${transpiled.optional ? ' = None' : ''}`;
  }

  private formatProperty(
    name: string,
    typeReference: transpile.TranspiledTypeReference,
  ): string {
    const tf = typeReference.toString({
      typeFormatter: (t) => t.name,
    });
    return `${name}: ${tf}`;
  }
}
