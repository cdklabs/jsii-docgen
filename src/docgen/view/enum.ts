import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { Transpile } from '../transpile/transpile';
import { EnumMember } from './enum-member';

export class Enum {
  private readonly members: EnumMember[];
  constructor(
    private readonly transpile: Transpile,
    private readonly enu: reflect.EnumType,
  ) {
    this.members = enu.members.map((em) => new EnumMember(transpile, em));
  }

  public render(): Markdown {
    const transpiled = this.transpile.enum(this.enu);
    const md = new Markdown({ header: { title: transpiled.name } });

    md.table([
      ['Name', 'Description'].map(Markdown.bold),
      ...this.members.map((member) => [member.linkedName, Markdown.sanitize(member.description)]),
    ]);
    md.split();

    if (this.enu.docs) {
      md.docs(this.enu.docs);
    }

    for (const m of this.members) {
      md.section(m.render());
    }

    return md;
  }
}
