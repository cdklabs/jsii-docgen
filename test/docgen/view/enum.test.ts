import * as reflect from 'jsii-reflect';
import { CSharpTranspile } from '../../../src/docgen/transpile/csharp';
import { JavaTranspile } from '../../../src/docgen/transpile/java';
import { PythonTranspile } from '../../../src/docgen/transpile/python';
import { TypeScriptTranspile } from '../../../src/docgen/transpile/typescript';
import { Enum } from '../../../src/docgen/view/enum';
import { Assemblies } from '../assemblies';

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;

describe('python', () => {
  const transpile = new PythonTranspile();
  test('snapshot', () => {
    const enu = new Enum(transpile, assembly.enums[0]);
    expect(enu.render().render()).toMatchSnapshot();
  });
});

describe('typescript', () => {
  const transpile = new TypeScriptTranspile();
  test('snapshot', () => {
    const enu = new Enum(transpile, assembly.enums[0]);
    expect(enu.render().render()).toMatchSnapshot();
  });
});

describe('java', () => {
  const transpile = new JavaTranspile();
  test('snapshot', () => {
    const enu = new Enum(transpile, assembly.enums[0]);
    expect(enu.render().render()).toMatchSnapshot();
  });
});

describe('csharp', () => {
  const transpile = new CSharpTranspile();
  test('snapshot', () => {
    const enu = new Enum(transpile, assembly.enums[0]);
    expect(enu.render().render()).toMatchSnapshot();
  });
});
