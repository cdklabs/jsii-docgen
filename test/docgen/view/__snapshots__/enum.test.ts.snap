// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "Amazon.CDK.AWS.ECR.TagMutability",
  "id": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "Amazon.CDK.AWS.ECR.TagMutability.MUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "Amazon.CDK.AWS.ECR.TagMutability.IMMUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
    },
  ],
}
`;

exports[`csharp snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.MUTABLE\\">MUTABLE</a></code> | allow image tags to be overwritten. |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\">IMMUTABLE</a></code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

The tag mutability setting for your repository.

# \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


# \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;

exports[`java snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "software.amazon.awscdk.services.ecr.TagMutability",
  "id": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "software.amazon.awscdk.services.ecr.TagMutability.MUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "software.amazon.awscdk.services.ecr.TagMutability.IMMUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
    },
  ],
}
`;

exports[`java snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.MUTABLE\\">MUTABLE</a></code> | allow image tags to be overwritten. |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\">IMMUTABLE</a></code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

The tag mutability setting for your repository.

# \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


# \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;

exports[`python snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "aws_cdk.aws_ecr.TagMutability",
  "id": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "aws_cdk.aws_ecr.TagMutability.MUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "aws_cdk.aws_ecr.TagMutability.IMMUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
    },
  ],
}
`;

exports[`python snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.MUTABLE\\">MUTABLE</a></code> | allow image tags to be overwritten. |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\">IMMUTABLE</a></code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

The tag mutability setting for your repository.

# \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


# \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;

exports[`typescript snapshot 1`] = `
Object {
  "displayName": "TagMutability",
  "docs": Object {
    "summary": "The tag mutability setting for your repository.",
  },
  "fqn": "@aws-cdk/aws-ecr.TagMutability",
  "id": "@aws-cdk/aws-ecr.TagMutability",
  "members": Array [
    Object {
      "displayName": "MUTABLE",
      "docs": Object {
        "summary": "allow image tags to be overwritten.",
      },
      "fqn": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.MUTABLE",
    },
    Object {
      "displayName": "IMMUTABLE",
      "docs": Object {
        "summary": "all image tags within the repository will be immutable which will prevent them from being overwritten.",
      },
      "fqn": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
      "id": "@aws-cdk/aws-ecr.TagMutability.IMMUTABLE",
    },
  ],
}
`;

exports[`typescript snapshot 2`] = `
" TagMutability <a name=\\"TagMutability\\" id=\\"@aws-cdk/aws-ecr.TagMutability\\"></a>

| **Name** | **Description** |
| --- | --- |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.MUTABLE\\">MUTABLE</a></code> | allow image tags to be overwritten. |
| <code><a href=\\"#@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\">IMMUTABLE</a></code> | all image tags within the repository will be immutable which will prevent them from being overwritten. |

---

The tag mutability setting for your repository.

# \`MUTABLE\` <a name=\\"MUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.MUTABLE\\"></a>

allow image tags to be overwritten.

---


# \`IMMUTABLE\` <a name=\\"IMMUTABLE\\" id=\\"@aws-cdk/aws-ecr.TagMutability.IMMUTABLE\\"></a>

all image tags within the repository will be immutable which will prevent them from being overwritten.

---

"
`;
